File: day22/KthLargest.java
import java.util.*;

public class KthLargest {

    public static int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll(); // remove smallest in heap
            }
        }
        return minHeap.peek(); // root is the kth largest
    }

    public static void main(String[] args) {
        Random rand = new Random();
        int length = rand.nextInt(21) + 10; // 10 to 30
        int[] nums = new int[length];
        for (int i = 0; i < length; i++) {
            nums[i] = rand.nextInt(101); // 0 to 100
        }

        int k = rand.nextInt(length) + 1; // 1 to length

        System.out.println("Input: " + Arrays.toString(nums));
        System.out.println("k: " + k);
        System.out.println(k + "th largest element: " + findKthLargest(nums, k));
    }
}

File: day22/kth_largest.py
import random
import heapq    # provides implementation of heapq algorithm

def kth_largest(nums, k):
    return heapq.nlargest(k, nums)[-1]

def generate_test_case():
    length = random.randint(10, 30)
    nums = [random.randint(0, 100) for _ in range(length)]
    k = random.randint(1, length)
    return nums, k

def main():
    nums, k = generate_test_case()
    print("Input:", nums)
    print("k:", k)
    print(f"{k}th largest element:", kth_largest(nums, k))

if __name__ == "__main__":
    main()
